apps/emqx_resource/src/emqx_resource_worker.erl

```diff
@@ -283,112 +283,63 @@ pick_cast(Id, Key, Query) ->
         ok
     end).
 
-do_resume(#{id := Id, name := Name} = Data) ->
+resume_from_blocked(Data) ->
+    #{name := Name} = Data,
     case inflight_get_first(Name) of
         empty ->
-            retry_queue(Data);
+            {next_state, running, Data};
         {Ref, FirstQuery} ->
             %% We retry msgs in inflight window sync, as if we send them
             %% async, they will be appended to the end of inflight window again.
-            retry_inflight_sync(Id, Ref, FirstQuery, Name, Data)
+            retry_inflight_sync(Ref, FirstQuery, Data)
     end.
 
-retry_queue(
+retry_inflight_sync(Ref, QueryOrBatch, Data0) ->
```

Comment:
I think it must still retry the inflight while blocked, otherwise we might get stuck if the request was lost and a response is never received.  So it would implement "at least once" delivery.  

Otherwise, we'd need to implement some kind of timeout for the inflight requests.  :thinking: